### ISSUE TITLE
Better control of Spring beans component scan

### ISSUE THREAD

**Issue by carlspring** [strongbox/strongbox#1522]
# Task Description

We have a real problem with Spring configuration classes and `@ComponentScan`. The problem is that all our modules share similar package names like `org.carlspring.strongbox.services` or `org.carlspring.strongbox.config` which can cause ambiguous behavior when using the `@ComponentScan` annotation. We can't say for sure what configuration will fetch beans from such packages and how to manage bean dependencies correctly.

We need to investigate our options and address this matter.

**Please, note:** This task requires very good understanding of the Spring framework.

# Tasks

The following tasks will need to be carried out:
* [ ] Analyze our configuration.
* [ ] Create a list of options for us.
* [ ] Sort this out.

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @sbespalov
  * @fuss86
  * @carlspring
  * @steve-todorov

**Comment by bono007** [comment]
@fuss86 
@carlspring 

### Analysis:

* Each module has no separation from other modules by package name. 
* Each module enables `@CopmonentScan` across a set of (usually wide) package names - which will bleed across modules

### Suggestion:

1. Disable component scanning in the main application class by switching from `@SpringBootApplication` to 
```java
@SpringBootConfiguration
@EnableAutoConfiguration
```
Right now there is really nothing it can scan as its package is nested down below the root package - but for clarity and expressiveness we should disable it.

2. Add a package name in each module to scope it from other modules. (ex: `org.carlspring.strongbox.config.DataServiceConfig` would move to `org.carlspring.strongbox.data.config.DataServiceConfig`)

3. Have each module produce top-level(s) (usable by other modules) `@Configuration` which in turn will `@ComponentScan` only in that modules newly scoped package. Any cross-module configurations needed can then be manually imported via `@Import` statement. Originally I would love to not use scanning at all but the amount of beans that would have to be specified currently is large. We can always do that later as this will move us in that direction.

**Comment by sbespalov** [comment]
@bono007 thanks for looking into it. Your suggestion is just like I suppose it should be, but this will force all our code to be moved inside new packages and it looks like epic refactoring, too epic probably. :)
Anyway this option is still an option and we should see @strongbox/core-developers thoughts

Another option is following:
1. Use `@ComponentScan` only within `CommonConfig` and import it where needed
2. Move  `@Configuration` classes under module specific package in all modules

**Comment by carlspring** [comment]
Would it be possible to just move the `@Configuration`-annotated classes under their own per-module-packages? Would this suffice...?

@fuss86 : Any thoughts on this?

**Comment by fuss86** [comment]
Re-package would be best, but as @sbespalov noticed, it is huge change.

Another alternative is to use `@Configuration`, `@Bean` and `constructor injection`. This would lead to better control of components creation. We could start from `strongbox-web-core` and go down module by module.

**Comment by fuss86** [comment]
And my proposal above should be done with `@Configuration` classes re-package as suggested above by predecessors :)

**Comment by carlspring** [comment]
Would it be sufficient to just move the `@Configuration` classes and do the `@Bean` + constructor injection at first in order to drop the `@ComponentScan` annotations?

**Comment by fuss86** [comment]
Yes, I believe so

**Comment by carlspring** [comment]
Okay, then I guess we're all on the same page. @bono007 , please feel free to proceed as discussed! :)

**Comment by bono007** [comment]
Sounds good. This will take a while but I will chip on it module/module over the week.

**Comment by bono007** [comment]
Ok so digging into this and starting to do this approach turns out to be more "epic" than a package refactoring. The reason is that we are basically going to replace every auto-configured `@Component` and `@Service` with a manually configured `@Bean` counterpart. This is a HUGE effort. Additionally, it will be more risky as switching to bean-level configuration as its a change to the entire surface area IMO.

**Comment by carlspring** [comment]
In this case, perhaps you could pick another task, while we discuss this internally with the @strongbox/core-developers team and weigh the pros and cons of doing this right now...?

**Comment by sbespalov** [comment]
also as an option we can put bean class names somewhere under module `META-INF` and make Spring to scan only classes that defined there.

**Comment by bono007** [comment]
Yep I think we should wait on this task until we upgrade to Spring 2.2 in which there is the [spring-context-indexer](https://docs.spring.io/spring/docs/5.1.2.RELEASE/spring-framework-reference/core.html#beans-scanning-index) in Spring 5.2 that will do this for us. 

In fact, this could be the "business motivator" for doing the upgrade.

**Comment by carlspring** [comment]
@sbespalov , @fuss86 , @ptirador ,

Perhaps we should indeed investigate [spring-context-indexer](https://docs.spring.io/spring/docs/5.1.2.RELEASE/spring-framework-reference/core.html#beans-scanning-index) first?

@sbespalov : We need follow-up tasks for upgrading to the latest version of Spring Security and Spring Boot (where your pull request has been merged). Could you please create these tasks? Thanks! :)

**Comment by carlspring** [comment]
This task is blocked by #1649 and can only be looked at once it is complete.
